### ISSUE TITLE
Provide route decorator similar to Grafana Beyla

### ISSUE THREAD

**Issue by titaneric** [grafana/alloy#1596]
### Request

Hi, I would like to request a new feature called [route decorator](https://grafana.com/docs/beyla/latest/configure/options/#routes-decorator) similar to Grafana Beyla.
It works like this (in Beyla config)

```
routes:
  patterns:
    - /user/{id}
    - /user/{id}/basket/{product}
```

and it would match `/user/123` and `/user/456` and it would include `/user/{id}` in the trace property.

Hope that there is a URL parsing function or a dedicated block called `matchRoute` in `log.process` component  to process the HTTP path like this.

Indeed, we have `regex` and `replace` block in `log.process` component. However, it's not powerful enough.

### Use case

Like I mentioned in #1592, I would to pre-process the performance resource metrics sent by Faro SDK. There is a `event_data` field in the `faro.receiver` log and it would have the full path of HTTP request. I want to remove any parameters, hostname, and any GET single resource (`/user/{id}`) before sending to Grafana Loki. By configuring these, I could transform the logql query into Prometheus metrics by Loki rulers.

**Comment by titaneric** [comment]
I notice that alloy do have [loki.template](https://grafana.com/docs/alloy/latest/reference/components/loki/loki.process/#stagetemplate-block), and it support [sprig URL parsing](https://masterminds.github.io/sprig/url.html). I think it could meet my need, and I would close this issue.

**Comment by titaneric** [comment]
Reopen here since I want to provide the following config to component or stage.
```
    - /user/{id}
    - /user/{id}/basket/{product}
```

**Comment by ptodev** [comment]
Hi, @titaneric! Do I understand correctly that `stage.template` can do the replacement you need, but you don't like the ergonomics? Is it because you need to replace more than one value (e.g. `id` and `product`), and you need to add multiple `stage.template` instances?

**Comment by titaneric** [comment]
Hi @ptodev, thanks for response. Basically, yes, and I think present alloy component could not handle multiple URL paths easily.

I think `stage.template` or `stage.replace` could help me to extract the path from the full URL. However, I want more than that. I want to replace any path to given *route decorator* to minimize the possibility of high cardinality.

Given the URL `https://example.com/user/123/wishList/456`,  and given imagined `stage.route_decorator` stage in `loki.process` components in alloy config.

```
loki.process "extract_path" {
  stage.replace {
    expression = "(?P<extracted_url>http\\S+)"
    replace = `{{  index (urlParse .Value) "path") }}`
  }
  stage.route_decorator {
    source= ""  
    decorator = [
      "/user/{id}",
      "/user/{id}/wishList/{product}"
    ]
  }
}
```

I hope this config could output `"/user/{id}/wishList/{product}` .

Hope that `stage.route_decorator` could replace the log line, and it could also replace the extracted values in `loki.process` component.
Perhaps a `route_decorator` function is another possible solution.

**Comment by titaneric** [comment]
There is no further response here, could I volunteer to implement this feature and raise PR?

**Comment by ptodev** [comment]
Hi @titaneric, apologies for the delayed response! IIUC, you'd like to replace the log line with a redacted URL? E.g. if the log is "a rq on /user/1234", then the output log will be simply "/user/{id}"? And then you'd like to query the number of "/user/{id}" logs, and generate a metric? I'm just not sure why you wouldn't simply use [stage.metrics](https://grafana.com/docs/alloy/latest/reference/components/loki/loki.process/#stagemetrics-block) to generate the metric directly in alloy?

**Comment by titaneric** [comment]
Yes, I have used `stage.metrics` to generate the the metrics that I want. However, since path like `/user/1234` is a relatively high cardinality, and it will cause the Prometheus overhead, and I want to preprocess the path in the Grafana Alloy. If I could provide the **redacted** URL template similar to Beyla, I could replace the path to the redacted one. I could easily query the Prometheus with path label, and it also don't leave the overhead to Prometheus.
